### PR TITLE
[SPARK-13647][SQL] also check if numeric value is within allowed range in _verify_type

### PR DESCRIPTION
**Pull request by cloud-fan**
## What changes were proposed in this pull request?

This PR makes the `_verify_type` in `types.py` more strict, also check if numeric value is within allowed range.
## How was this patch tested?

newly added doc test.

### CODE REVIEW

File: python/pyspark/sql/types.py [hunk 1]
```diff
@@ -1137,7 +1148,28 @@ def _verify_type(obj, dataType):
...
+    elif isinstance(dataType, IntegerType):
+        if obj < -2147483648 or obj > 2147483647:
+            raise ValueError("object of IntegerType out of range, got: %s" % obj)
+
+    elif isinstance(dataType, FloatType):
+        from math import isinf
+        from struct import pack, unpack
+
+        if not isinf(obj):
+            f = unpack("f", pack("f", obj))[0]
```
Comment (cloud-fan) [inline python/pyspark/sql/types.py]: Here I try to convert the `obj` to binary string as a `float32`, and then convert it back. If it becomes an `inf`, then it means the `obj` can not fit in `float32`.

Is there a better way to check if a python float can fit in `float32`?

**Comment by cloud-fan** [comment]
cc @yhuai @davies

File: python/pyspark/sql/types.py (hunk 1 above)
Comment (davies) [inline python/pyspark/sql/types.py]: I'm worried that this will be super slow, I think we don't need this.

If overflow happens, it will become infinity.

**Comment by davies** [comment]
LGTM

**Comment by davies** [comment]
Merging this into master, thanks!

**Pull request closed without merging**